Backpack/HorizontalNavigation/Classes/BPKHorizontalNavigation.swift

```diff
@@ -0,0 +1,520 @@
+/*
+ * Backpack - Skyscanner's Design System
+ *
+ * Copyright 2018-2021 Skyscanner Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+import UIKit
+
+
+public protocol BPKHorizontalNavigationSize {
+    static var fontStyle: BPKFontStyle { get }
+
+    static var verticalSpacing: CGFloat { get }
+}
+
+public enum BPKHorizontalNavigationSizeDefault: BPKHorizontalNavigationSize {
```

Comment:
The reason for enum is exactly because we only use this at the type level though. We don't want to create instance of it which an `enum` without cases nicely supports

Reply:
Guess same as above, why do we want it to be type level ?

Reply:
Because the goal is to achieve a type safe API that doesn't allow misuse, in particular it should be impossible to use a small icon with the default size of the horizontal navigation and vice versa.

This is the case now because of the `where` constraint in this definition

```swift
        public static func textAndIcon<
            S,
            Icon: BPKHorizontalNavigationOptionIcon
        >(name: String, icon: Icon, tag: Int = 0, showNotificationDot: Bool = false) -> AnyOption<S>
        where Icon.Size == S {
```

This prevents you from writing the following code:

```swift
let hoirzontalNav = BPKHorizontalNavigation<BPKHorizontalNavigationSizeDefault>(options: [
  .textAndIcon("Hotels", icon: BPKIconNameSmall.hotels, tag: 0)
], selectedItemIndex: 0)
```

In the previous Objective-c version we had a `size` property but that doesn't work as are migrating the icon API to one where not all icons are available at all sizes.

Reply:
Hm, Ok, I see your point!
I think this is a use case that makes sense for the main app, but since this is an open source lib, maybe there are some cases where clients would like a bit more control over what to draw, even if it's now what our designers state, like using a big image with a small text.
Either way, I got why it's been done this way now!